File: Data Structure/Linked List/Medium/LET_no1429_First Unique Number.py
class Node:
    def __init__(self, val):
        self.val = val
        self.prev, self.next = None, None

class FirstUnique:

    def __init__(self, nums: List[int]):
        self.dummy = Node(-1)
        self.tail = self.dummy
        self.memo = {}
        self.dups = set()
        for num in nums:
            self.add(num)

    def showFirstUnique(self) -> int:
        if self.dummy.next:
            return self.dummy.next.val
        return -1

    def add(self, value: int) -> None:
        if value in self.dups:
            return
        if value in self.memo:
            self.popTarget(value)
            return
        
        cur = Node(value)
        self.memo[value] = cur
        self.tail.next = cur
        cur.prev = self.tail
        cur.next = None
        self.tail = cur
    
    def popTarget(self, value):
        print(value, self.tail.val)
        cur = self.memo[value]
        del self.memo[value]
        self.dups.add(value)
        if cur is self.tail:
            self.tail = cur.prev
            self.tail.next = None
            return
        
        p, n = cur.prev, cur.next
        p.next = n
        n.prev = p
        


# Your FirstUnique object will be instantiated and called as such:
# obj = FirstUnique(nums)
# param_1 = obj.showFirstUnique()
# obj.add(value)

File: Data Structure/Linked List/Medium/LET_no1429_First Unique Number.cpp
class FirstUnique {
private:
    list<int> q;
    unordered_map<int, list<int>::iterator> key2itr;
    unordered_set<int> duplicates;
public:
    FirstUnique(vector<int>& nums) {
        for (int num : nums)
            add(num);
    }
    
    int showFirstUnique() {
        return !key2itr.empty() ? q.front() : -1;
    }
    
    void add(int value) {
        if (duplicates.count(value))
            return;
        else if (key2itr.count(value)) {
            duplicates.insert(value);
            auto it = key2itr[value];
            q.erase(it);
            key2itr.erase(value);
        }
        else {
            q.push_back(value);
            key2itr[value] = --q.end();
        }
    }
};

/**
 * Your FirstUnique object will be instantiated and called as such:
 * FirstUnique* obj = new FirstUnique(nums);
 * int param_1 = obj->showFirstUnique();
 * obj->add(value);
 */